### PR TITLE
clustering: new parallelization implementation

### PR DESCRIPTION
**Pull request by MSusik**
* Gets rid of Joblib

* Removes feature of passing negative value of n_jobs.

Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>

### CODE REVIEW

File: beard/clustering/blocking.py [hunk 1]
```diff
@@ -159,33 +183,48 @@ def _blocks(self, X, y, blocks):
...
-                   b, X_mask, y_mask, clusterer in self._blocks(X, y, blocks)))
-
-        for b, clusterer in results:
-            self.clusterers_[b] = clusterer
+        processes = []
+        # Here the blocks will be passed to subprocesses
+        data_queue = mp.Queue()
+        # Here the results will be passed back
+        result_queue = mp.Queue()
+        for x in range(0, self.n_jobs):
```
Comment (glouppe) [inline beard/clustering/blocking.py]: `range(self.n_jobs)`

File: beard/clustering/blocking.py [hunk 2]
```diff
@@ -159,33 +183,48 @@ def _blocks(self, X, y, blocks):
...
+            processes.append(mp.Process(target=_parallel_fit, args=(self.fit_,
+                             self.partial_fit_, self.base_estimator,
+                             self.verbose, data_queue, result_queue)))
+            processes[-1].start()
+
+        # First 16 blocks are sent into the queue without waiting for the
+        # results. This variable is a counter that takes care of this.
+        presend = 1
```
Comment (glouppe) [inline beard/clustering/blocking.py]: I dont understand from the code how the first *16* are sent into the queue. Where does this number come from?

File: beard/clustering/blocking.py [hunk 3]
```diff
@@ -159,33 +183,48 @@ def _blocks(self, X, y, blocks):
...
+            else:
+                presend += 1
+            if self.partial_fit_:
+                if block[0] in self.clusterers_:
+                    data_queue.put(('middle', block, self.clusterers_[b]))
+                    continue
+            data_queue.put(('middle', block, None))
+
+        # Get the last results and tell the subprocesses to finish
+        for x in range(0, self.n_jobs):
```
Comment (glouppe) [inline beard/clustering/blocking.py]: `range(self.n_jobs)`

File: beard/clustering/blocking.py [hunk 4]
```diff
@@ -159,33 +183,48 @@ def _blocks(self, X, y, blocks):
...
+                             self.partial_fit_, self.base_estimator,
+                             self.verbose, data_queue, result_queue)))
+            processes[-1].start()
+
+        # First 16 blocks are sent into the queue without waiting for the
+        # results. This variable is a counter that takes care of this.
+        presend = 1
+
+        for block in self._blocks(X, y, blocks):
+            if presend > self.n_jobs:
```
Comment (glouppe) [inline beard/clustering/blocking.py]: From a rigorous loop invariant point of view, I would initialize `presend` to `0` and then check whether `presend >= self.n_jobs` instead. Or better, I would put this check after the `data_queue.put` call below (using a `>` then).

**Comment by glouppe** [comment]
Besides my nitpicks, this looks good to me. Thanks! Two questions though:

- Does this solve the blocking issue we had?
- Have you checked this works fine both for Python 2.7 and 3.4?

**Comment by MSusik** [comment]
> Does this solve the blocking issue we had?

Yes, just a one small improvement is needed, as this implementation can hang.

> Have you checked this works fine both for Python 2.7 and 3.4?

Yes.

**Comment by glouppe** [comment]
Great then, +1 for merge once my comments are fixed.

**Comment by glouppe** [comment]
CC: @ogrisel Just to let you know we have add hanging issues with joblib -- processes are all stalling, CPU usage goes down to 0, and then nothing more happens. Do you know where this could be coming from? These are very difficult to reproduce and seem to appear at random... Directly using `multiprocessing` solves our immediate problem, but it would be nice if joblib could be used again.

**Comment by MSusik** [comment]
Note that the approach is completely different as joblib spawns a process for every block of data, while I keep a pool of processes that run all the time.

**Comment by ogrisel** [comment]
I want to make it possible to reuse joblib pools too across several consecutive calls to `Parallel.__call__`. However this is probably not the cause of the hanging.

Which implementation of BLAS do you use when you observe the hanging? anaconda's MKL? OSX's builtin accelerate?

**Comment by ogrisel** [comment]
> Note that the approach is completely different as joblib spawns a process for every block of data, while I keep a pool of processes that run all the time.

Note: joblib spawns a pool with a fixed number of worker process per call to Parallel (using a multiprocessing `Pool` instance using the the `apply_async` under the hood). But you can pass many blocks of data and the number of workers stay constants.

**Comment by ogrisel** [comment]
BTW, could you also tell be if you observe the hanging behavior when enabling the `forkserver` start method under Python 3.4+? To try that you need to modify the main block of the main script that starts your Python program with:

```
import multiprocessing as mp
# import your modules here

if __name__ == '__main__':
    mp.set_start_method('forkserver')
    # call your code here
```

**Comment by MSusik** [comment]
> Which implementation of BLAS do you use when you observe the hanging? anaconda's MKL?

We observed the error when using anaconda's MKL and when working outside of anaconda. For example here is a numpy's config for the run without anaconda:

```
blas_info:
    libraries = ['blas']
    library_dirs = ['/usr/lib']
    language = f77
lapack_info:
    libraries = ['lapack']
    library_dirs = ['/usr/lib']
    language = f77
atlas_threads_info:
  NOT AVAILABLE
blas_opt_info:
    libraries = ['blas']
    library_dirs = ['/usr/lib']
    language = f77
    define_macros = [('NO_ATLAS_INFO', 1)]
atlas_blas_threads_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
lapack_opt_info:
    libraries = ['lapack', 'blas']
    library_dirs = ['/usr/lib']
    language = f77
    define_macros = [('NO_ATLAS_INFO', 1)]
atlas_info:
  NOT AVAILABLE
lapack_mkl_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
mkl_info:
  NOT AVAILABLE
```

**Comment by ogrisel** [comment]
That's weird. Do you use OpenMP-based libraries or compiled extensions (e.g. Cython `prange` constructs)?

Would be great to provide a standalone joblib snippet that reproduce the freeze so that I can try to debug.

**Pull request merged**